### DISCUSSION TITLE
Reusing entities across multiple projects

### DISCUSSION THREAD

**Discussion by iamchathu** [mikro-orm/mikro-orm#2111] [Q&A]
We got a monorepo setup with multiple projects with same like database structure.

I was able to create a core Entities, most of the time we can use the core but there are cases where entities have additional fields. Plan was to extend the core inside the project and use that subclass as the entity.

But since the Entity relationships it uses the class in the core so when we put the subclass to the entities it get an error.

We are currently using `typeorm` and due to this issue we are currently seeking alternative to handle this.

Will I be able to use `mikro-orm` for this use case?

## EDIT:

Let say we got a project where there is a table for items which has a one to many relationship with categories. we put these two entities in a reusable core package in a mono-repo

There is also another project in same mono-repo which has some additional fields in items table.

When I try to load the entity, there will be issue since Category entity will refer the super class version of Item in core package.

How can we address this kind of scenario?

**Comment by B4nan** [comment]
I really don't understand what you are asking for.

**Reply by iamchathu** [reply]
Updated the question

**Comment by iamchathu** [comment]
@B4nan  Let get an example.

Let say we got a project where there is a table for `items` which has a one to many relationship with `categories`. we put these two entities in a reusable core package in a mono-repo

There is also another project in same mono-repo which has some additional fields in `items` table.

When I try to load the entity, there will be issue since `Category` entity will refer the super class version of `Item` in core package.

How can we address this kind of scenario?

**Reply by B4nan** [reply]
It is best if you try to prototype it yourself. But if you had issues to model this scenario with typeorm, I can imagine it will be problematic here too - but maybe not, from this I am still not fully sure what you are up to, so try it to verify. 

Will be happy to look into actual example repo and try to help with that.

**Reply by iamchathu** [reply]
@B4nan I added you to minimal sample repo

**Reply by B4nan** [reply]
please add some readme with what you want me to look for.

also note that you will need to declare the orm packages as peer depenencies, this will almost certainly not work as expected due to multiple `core` package installations, resulting in instanceof checks not working anywhere

**Reply by iamchathu** [reply]
Just added the readme. It does work currently for the project-1

**Reply by iamchathu** [reply]
@B4nan did you had time to look in to this?

**Reply by B4nan** [reply]
nope, and i dont even see any link to the repository you mentioned

**Reply by iamchathu** [reply]
@B4nan https://github.com/iamchathu/monorepo-sample

**Reply by B4nan** [reply]
you can't just extend a regular entity, you either use STI (to have single db table) or a shared (abstract) base entity

**Reply by iamchathu** [reply]
How to handle the relationships (Current issue with it)?

**Reply by B4nan** [reply]
as said above, you should not have multiple core packages installed (which you will with current setup), otherwise decorators cant work, as there will be multiple static registers for the decorator metadata. 

i believe you should define the orm as peer dependency here (both packages)

https://github.com/iamchathu/monorepo-sample/blob/master/packages/core/package.json#L17

**Reply by iamchathu** [reply]
I did it now. When I run project-2 it gives

```bash
[class Category]
[discovery] ORM entity discovery started, using ReflectMetadataProvider
[discovery] - processing entity Item
[discovery] - processing entity Category
(node:88984) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'name' of undefined
```

**Reply by B4nan** [reply]
the thing is, that all entities need to be discovered, including the base entities. in your project 2 the original `Item` entity is not, and you end up just with a single property in the metadata. it is forbidden to have multiple entities with same name, which you basically have.
